pmlc/dialect/abi/ir/example.mlir

```diff
@@ -0,0 +1,349 @@
+// Before lowering to ABI:
+
+module @dot {
+  func @main(%arg0: memref<8x16xf32>, %arg1: memref<16x32xf32>, %arg2: memref<8x32xf32>) {
+    %cst = constant 0.000000e+00 : f32
+    %c0 = constant 0 : index
+    %c1 = constant 1 : index
+    %c2 = constant 2 : index
+    %c3 = constant 3 : index
+    %c5 = constant 5 : index
+    %c6 = constant 6 : index
+    %c8 = constant 8 : index
+    %c7 = constant 7 : index
+    %c4 = constant 4 : index
+    omp.parallel num_threads(%c4 : index) default(shared) {
+      %1 = call @plaidml_rt_thread_num() : () -> index
+      br ^bb1(%c0 : index)
+    ^bb1(%2: index):  // 2 preds: ^bb0, ^bb2
+      %3 = cmpi "slt", %2, %c8 : index
+      cond_br %3, ^bb2, ^bb3
+    ^bb2:  // pred: ^bb1
+      %4 = muli %1, %c8 : index
+      store %cst, %arg2[%2, %4] : memref<8x32xf32>
+      %5 = addi %4, %c1 : index
+      store %cst, %arg2[%2, %5] : memref<8x32xf32>
+      %6 = addi %4, %c2 : index
+      store %cst, %arg2[%2, %6] : memref<8x32xf32>
+      %7 = addi %4, %c3 : index
+      store %cst, %arg2[%2, %7] : memref<8x32xf32>
+      %8 = addi %4, %c4 : index
+      store %cst, %arg2[%2, %8] : memref<8x32xf32>
+      %9 = addi %4, %c5 : index
+      store %cst, %arg2[%2, %9] : memref<8x32xf32>
+      %10 = addi %4, %c6 : index
+      store %cst, %arg2[%2, %10] : memref<8x32xf32>
+      %11 = addi %4, %c7 : index
+      store %cst, %arg2[%2, %11] : memref<8x32xf32>
+      %12 = addi %2, %c1 : index
+      br ^bb1(%12 : index)
+    ^bb3:  // pred: ^bb1
+      omp.terminator
+    }
+    %0 = xsmm.gemm.dispatch.f32 [8, 32, 16], [16, 32, 32]
+    xsmm.gemm.invoke.f32 %0, %arg2[%c0, %c0] = %arg0[%c0, %c0], %arg1[%c0, %c0] : (memref<8x16xf32>, memref<16x32xf32>) -> memref<8x32xf32>
+    return
+  }
+  func @plaidml_rt_thread_num() -> index
+}
+
+// After lowering to ABI:
+
+module @dot {
+  "abi.loop"() ( {
```

Comment:
Can you make a custom printer/parser for this op? It's pretty hard to read in generic form.

Reply:
Agree, checkout the Declarative Assembly Form bit in: 'https://mlir.llvm.org/docs/OpDefinitions/'. It's actually very easy, even for things with regions now.